File: tests/core/contractors/codac2_tests_CtcFixpoint.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *


class CtcCustom(Ctc):

  def __init__(self):
    Ctc.__init__(self, 2)

  def contract(self, x):
    if x[0].rad() > 1.:
      return IntervalVector(x.mid()).inflate(0.5*x.rad())
    return x


class TestCtcFixpoint(unittest.TestCase):
  
  def tests_CtcFixpoint(self):

    ctc_custom = CtcCustom()
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-500,500],[-500,500]]))
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-250,250],[-250,250]]))
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-125,125],[-125,125]]))

    ctc_fixed = CtcFixpoint(ctc_custom, 0.)
    
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x,1e-1) == IntervalVector([[-1,1],[-1,1]]))
    x = IntervalVector([[-500,500],[-500,500]])
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x,1e-1) == IntervalVector([[-1,1],[-1,1]]))

    ctc_fixed = CtcFixpoint(ctc_custom, 0.8)
    
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-500,500],[-500,500]]))
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-250,250],[-250,250]]))

    ctc_fixed = CtcFixpoint(ctc_custom, 1.)
    
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-500,500],[-500,500]]))
    ctc_fixed.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-250,250],[-250,250]]))


if __name__ ==  '__main__':
  unittest.main()

File: tests/core/contractors/codac2_tests_CtcFixpoint.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_CtcFixpoint.h>
#include <codac2_Approx.h>

using namespace std;
using namespace codac2;

class CtcCustom : public Ctc<CtcCustom,IntervalVector>
{
  public:

    CtcCustom()
      : Ctc<CtcCustom,IntervalVector>(2)
    { }

    void contract(IntervalVector& x) const
    {
      if(x[0].rad() > 1.)
        x = IntervalVector(x.mid()).inflate(0.5*x.rad());
    }
};

TEST_CASE("CtcFixpoint")
{
  CtcCustom ctc_custom;
  IntervalVector x({{-1000,1000},{-1000,1000}});
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-500,500},{-500,500}}));
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-250,250},{-250,250}}));
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-125,125},{-125,125}}));

  {
    CtcFixpoint ctc_fixed(ctc_custom, 0.);
    
    x = IntervalVector({{-1000,1000},{-1000,1000}});
    ctc_fixed.contract(x);
    CHECK(Approx(x,1e-1) == IntervalVector({{-1,1},{-1,1}}));
    x = IntervalVector({{-500,500},{-500,500}});
    ctc_fixed.contract(x);
    CHECK(Approx(x,1e-1) == IntervalVector({{-1,1},{-1,1}}));
  }

  {
    CtcFixpoint ctc_fixed(ctc_custom, 0.8);
    
    x = IntervalVector({{-1000,1000},{-1000,1000}});
    ctc_fixed.contract(x);
    CHECK(Approx(x) == IntervalVector({{-500,500},{-500,500}}));
    ctc_fixed.contract(x);
    CHECK(Approx(x) == IntervalVector({{-250,250},{-250,250}}));
  }

  {
    CtcFixpoint ctc_fixed(ctc_custom, 1.);
    
    x = IntervalVector({{-1000,1000},{-1000,1000}});
    ctc_fixed.contract(x);
    CHECK(Approx(x) == IntervalVector({{-500,500},{-500,500}}));
    ctc_fixed.contract(x);
    CHECK(Approx(x) == IntervalVector({{-250,250},{-250,250}}));
  }
}